packages/flutter/lib/src/rendering/sliver_list.dart

```diff
@@ -169,20 +169,22 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
     // If the scroll offset is at zero, we should make sure we are
     // actually at the beginning of the list.
     if (scrollOffset < precisionErrorTolerance) {
-      if (indexOf(firstChild) > 0) {
+      while (indexOf(firstChild) > 0) {
         final double earliestScrollOffset = childScrollOffset(firstChild);
-        // We correct one child at a time. If there are more children before
-        // the earliestUsefulChild, we will correct it once the scroll offset
-        // reach zero again.
-        earliestUsefulChild = insertAndLayoutLeadingChild(childConstraints, parentUsesSize: true);
-        assert(earliestUsefulChild != null);
         final double firstChildScrollOffset = earliestScrollOffset - paintExtentOf(firstChild);
```

Comment:
This does not look right, the `earliestScrollOffset` is a the beginning of the first child already, we want to take `earliestScrollOffset` and minus the paint extent of child before the first child. I think one thing that is confusing of the original code is that line 177 ` earliestUsefulChild = insertAndLayoutLeadingChild(childConstraints, parentUsesSize: true);` will actually update the firstChild. So the firstChild in line 173 and line 179 are actually different.